tests/Form/DataTransformerResolverTest.php

```diff
@@ -0,0 +1,221 @@
+
+    /**
+     * @var FieldDescriptionInterface
+     */
+    private $modelManager;
+
+    protected function setUp(): void
+    {
+        $this->fieldDescription = $this->prophesize(FieldDescriptionInterface::class);
+        $this->modelManager = $this->prophesize(ModelManagerInterface::class);
+        $this->resolver = new DataTransformerResolver();
+    }
+
+    public function testFailedResolve(): void
+    {
+        $this->assertNull($this->resolve());
+    }
+
+    public function provideFieldTypes(): array
+    {
+        return [
+            ['foo'],
+            // override predefined transformers
+            ['date'],
+            ['boolean'],
+            ['choice'],
+        ];
+    }
+
+    /**
+     * @dataProvider provideFieldTypes
+     */
+    public function testResolveCustomDataTransformer(string $fieldType): void
+    {
+        $customDataTransformer = new CallbackTransformer(static function ($value): string {
+            return (string) (int) $value;
+        }, static function ($value): bool {
+            return filter_var($value, FILTER_VALIDATE_BOOLEAN);
+        });
+        $this->fieldDescription->getOption('data_transformer')->willReturn($customDataTransformer);
+        $this->fieldDescription->getType()->willReturn($fieldType);
+
+        $dataTransformer = $this->resolve();
+
+        $this->assertInstanceOf(DataTransformerInterface::class, $dataTransformer);
+        $this->assertSame($customDataTransformer, $dataTransformer);
+    }
+
+    public function getTimeZones(): iterable
+    {
+        $default = new \DateTimeZone(date_default_timezone_get());
+        $custom = new \DateTimeZone('Europe/Rome');
+
+        return [
+            'empty timezone' => [null, $default],
+            'disabled timezone' => [false, $default],
+            'default timezone by name' => [$default->getName(), $default],
+            'default timezone by object' => [$default, $default],
+            'custom timezone by name' => [$custom->getName(), $custom],
+            'custom timezone by object' => [$custom, $custom],
+        ];
+    }
+
+    /**
+     * @dataProvider getTimeZones
+     */
+    public function testResolveDateDataTransformer($timezone, \DateTimeZone $expectedTimezone): void
+    {
+        $this->fieldDescription->getOption('data_transformer')->willReturn(null);
+        $this->fieldDescription->getOption('timezone')->willReturn($timezone);
+        $this->fieldDescription->getType()->willReturn('date');
+
+        $dataTransformer = $this->resolve();
+
+        $this->assertInstanceOf(DateTimeToStringTransformer::class, $dataTransformer);
+
+        $value = '2020-12-12';
+        $defaultTimezone = new \DateTimeZone(date_default_timezone_get());
+        $expectedDate = new \DateTime($value, $expectedTimezone);
+        $expectedDate->setTimezone($defaultTimezone);
+
+        $resultDate = $dataTransformer->reverseTransform($value);
+
+        $this->assertInstanceOf(\DateTime::class, $resultDate);
```

Comment:
```suggestion
        $this->assertInstanceOf(\DateTimeInterface::class, $resultDate);
```

Reply:
Now the transformer always returns the `\DateTime`. If we expect the interface in tests, we may not notice how we break BC by replacing `\DateTime` with `\DateTimeImmutable` in the transformer.